### ISSUE TITLE
opencl implementation

### ISSUE THREAD

**Issue by nihui** [Tencent/ncnn#132]
to be updated ...

**Comment by nihui** [comment]
design notes

changes in Mat
new Mat member to indicate host memory or device memory
new Mat member to hold cl_mem
allocate memory use CL_MEM_ALLOC_HOST_PTR
new Mat member function Mat::sync() for syncing memory (map + copy + wait)
opencl routine for Mat::fill()
opencl routine for Mat::clone()

changes in Layer
new Layer member to indicate support_opencl
new opencl subdirectory to include all opencl kernel source
glue layer source call clEnqueueNDRangeKernel

changes in Net
new Net member function Net::compile() to build opencl program
Net::forward_layer() accept extra argument cl_command_queue, opencl_mode, may wrap all runtime Extractor params into one structure
Net::forward_layer() check layer support_opencl and call sync() if needed

changes in Extractor
new Extractor member to hold cl_command_queue
new Extractor member function Extractor::set_opencl_mode(bool) for switching inference device
Extractor::input() call sync() to upload if needed
Extractor::extract() call sync() to download if needed

changes in global scope
new routine for initialize opencl device
new routine for get default device context
new storage class for compiling all kernel and program, cache them

changes in build system
cmake option NCNN_OPENCL

**Comment by nihui** [comment]
opencl branch
https://github.com/Tencent/ncnn/tree/opencl

**Comment by nihui** [comment]
we prefer vulkan to opencl for gpu inference.
ref #577

**Comment by xiaofengren** [comment]
may I know whether it can support OpenCL 1.2? On my arm platform, it only find OpenCL 1.2. 
Found OpenCL: /usr/lib/libOpenCL.so (found version "1.2")
When I build ncnn with OpenCL, it will report a lot of library unmatched error, like as followings:

../src/libncnn.a(modelbin.cpp.o): In function `ncnn::ModelBin::load(int, int, int, int) const':
modelbin.cpp:(.text+0x1354): undefined reference to `clReleaseMemObject'
modelbin.cpp:(.text+0x15dc): undefined reference to `clReleaseMemObject'
modelbin.cpp:(.text+0x172c): undefined reference to `clReleaseMemObject'
modelbin.cpp:(.text+0x1748): undefined reference to `clReleaseMemObject'
modelbin.cpp:(.text+0x17c0): undefined reference to `clCreateBuffer'
modelbin.cpp:(.text+0x1858): undefined reference to `clReleaseMemObject'
modelbin.cpp:(.text+0x18c8): undefined reference to `clCreateBuffer'
../src/libncnn.a(paramdict.cpp.o): In function `ncnn::ParamDict::set(int, ncnn::Mat const&)':
paramdict.cpp:(.text+0x248): undefined reference to `clReleaseMemObject'
../src/libncnn.a(paramdict.cpp.o): In function `ncnn::ParamDict::clear()':
paramdict.cpp:(.text+0x388): undefined reference to `clReleaseMemObject'
paramdict.cpp:(.text+0x3ac): undefined reference to `clReleaseMemObject'
../src/libncnn.a(paramdict.cpp.o): In function `ncnn::ParamDict::load_param(_IO_FILE*)':
paramdict.cpp:(.text+0x778): undefined reference to `clReleaseMemObject'
paramdict.cpp:(.text+0x7e4): undefined reference to `clCreateBuffer'
../src/libncnn.a(paramdict.cpp.o): In function `ncnn::ParamDict::load_param_bin(_IO_FILE*)':
paramdict.cpp:(.text+0xa98): undefined reference to `clCreateBuffer'
paramdict.cpp:(.text+0xad4): undefined reference to `clReleaseMemObject'
../src/libncnn.a(paramdict.cpp.o): In function `ncnn::ParamDict::load_param(unsigned char const*&)':
paramdict.cpp:(.text+0xcb0): undefined reference to `clCreateBuffer'
paramdict.cpp:(.text+0xd04): undefined reference to `clReleaseMemObject'

The GPU what I used is not Mali GPU, so can it be supported?
